internal/application/application.go

```diff
@@ -17,161 +17,113 @@ limitations under the License.
 package application
 
 import (
-	"bytes"
-	"errors"
+	"crypto/sha1"
 	"fmt"
 	"path/filepath"
 	"strings"
 	"unicode"
 
 	redskyappsv1alpha1 "github.com/thestormforge/optimize-controller/api/apps/v1alpha1"
-	"github.com/thestormforge/optimize-controller/internal/scan"
-	"sigs.k8s.io/kustomize/api/filesys"
-	"sigs.k8s.io/kustomize/api/provider"
-	"sigs.k8s.io/kustomize/api/resmap"
-	"sigs.k8s.io/kustomize/kyaml/kio"
 	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
 )
 
-// FilterScenarios retains only the named scenario on the supplied application. Removing unused
-// scenarios may be useful for some types of application operations. If the requested scenario
-// cannot be found, an error is returned.
-func FilterScenarios(app *redskyappsv1alpha1.Application, scenario string) error {
-	if scenario == "" {
-		if len(app.Scenarios) > 1 {
-			names := make([]string, 0, len(app.Scenarios))
-			for _, s := range app.Scenarios {
-				names = append(names, s.Name)
-			}
-			return fmt.Errorf("scenario is required (should be one of '%s')", strings.Join(names, "', '"))
-		}
-		return nil
+// GetScenario returns the named scenario from the application, if possible.
+func GetScenario(app *redskyappsv1alpha1.Application, scenario string) (*redskyappsv1alpha1.Scenario, error) {
+	if scenario == "" && len(app.Scenarios) == 1 {
+		return &app.Scenarios[0], nil
 	}
 
-	switch len(app.Scenarios) {
-
-	case 0:
-		return fmt.Errorf("unknown scenario '%s' (application has no scenarios defined)", scenario)
-
-	case 1:
-		if cleanName(app.Scenarios[0].Name) == scenario {
-			return nil
-		}
-		return fmt.Errorf("unknown scenario '%s' (must be '%s')", scenario, app.Scenarios[0].Name)
+	if len(app.Scenarios) == 0 {
+		return nil, nil
+	}
 
-	default:
-		names := make([]string, 0, len(app.Scenarios))
-		for i := range app.Scenarios {
-			if cleanName(app.Scenarios[i].Name) == scenario {
-				// Only keep the requested scenario
-				app.Scenarios = app.Scenarios[i : i+1]
-				return nil
+	cleanScenario := cleanName(scenario)
+	var result *redskyappsv1alpha1.Scenario
+	var names []string
+	for i := range app.Scenarios {
+		names = append(names, app.Scenarios[i].Name)
+		if cleanName(app.Scenarios[i].Name) == cleanScenario {
+			if result != nil {
+				return nil, fmt.Errorf("scenario name %q is ambiguous", scenario)
 			}
-			names = append(names, app.Scenarios[i].Name)
+			result = &app.Scenarios[i]
 		}
-		return fmt.Errorf("unknown scenario '%s' (should be one of '%s')", scenario, strings.Join(names, "', '"))
+	}
```

Comment:
thoughts?
```suggestion
if cleanName(app.Scenarios[i].Name) != cleanScenario {
  continue
}

if result != nil {
  return nil, fmt.Errorf("scenario name %q is ambiguous", scenario)
}

result = &app.Scenarios[i]
names = append(names, app.Scenarios[i].Name)
```

Reply:
Almost, the name appending needs to be before the continue check to ensure we get the full list for error messages.